docs/misc/new_design.md

```diff
@@ -0,0 +1,169 @@
+
+That's it. We don't need concrete types or computibility to reason about them. The semantic can end with their names and we don't need to know more underneath.
+
+This is the key insight one must have before formalizing a piece of mathematics.
+
+## Compatibility with mathlib
+
+A geometric product "contains" an inner product and an exterior product (or wedge product).
+
+They're established mathematic. There's already inner product space in mathlib and there would definitely be Grassmann Algebra in mathlib one day. And obviously we can't escape the general Clifford Algebra too.
+
+We must maintain compatibility with them, at least don't conflict with their existence.
+
+It's not a duplication. The development of mathlib is driven by professional mathematitians, they struggle for math generity but we're working at an abstraction level close to applications.
+
+We want to stand on such shoulders but we want good capsulation so that we don't want to worry about more abstract concepts and some too general cases. This would definitely leak, it might not be obvious in definitions but it would be very clear when you're proving something. For example, you have to deal with lattice and finsupp when proving things about linear algebra, and that should not be.
+
+## Operators
+
+We start with has_* that have absolutely no axioms with them, they don't have any properties and any behaviors. It's almost just a name. And we associate notations with them.
+
+Just like in https://github.com/leanprover-community/lean/blob/master/library/init/core.lean#L321 
+
+```lean
+class has_mul      (α : Type u) := (mul : α → α → α)
+```
+
+we could just:
+
+```lean
+class has_wedge (α : Type u) := (wedge : α → α → α)
+class has_inner (α : Type u) := (inner : α → α → α)
+```
+
+The latter presumably is already in mathlib and has a lot of structures and properties associated with it. We'll deal with that later.
+
+As for geometric product, I'm thinking about the following short name instead of `geomprod`, `gp` etc.:
+
+```
+class has_gmul (α : Type u) := (gmul : α → α → α)
+```
+
+## Notations
+
+Now we assoc notations with them:
+
+```lean
+-- \circ
+infix ∘ := has_gmul.gmul
+-- \wedge
+infix ∧ := has_wedge.wedge
+-- \cdot
+infix ⋅ := has_inner.inner
+```
+
+We'll always use `local` notations waiting for to be `open_locale`ed. I expect conflicts with notations in mathlib and using this to avoid the conflicts as long as possible.
+
+> TODO: describe how to use them even when there's a confliction.
+
+## Defining without defining
+
+There're really millions of products defined in GA and they're based on geometric product. But the definition might not be the most effient one or the most intuitive one.
+
+So instead of using `def`, we should make these products just a product with a Prop assering that they're equal to the definition based on gp and let the implementation prove it when intantiate the instance.
```

Comment:
```suggestion
So instead of using `def`, we should make these products just a product with a Prop asserting that they're equal to the definition based on gp and let the implementation prove it when intantiate the instance.
```